book/testing_code.md

```diff
@@ -240,35 +240,104 @@ Having one test class per function/class that you are testing will make it clear
 There are multiple test files for the `math.R` module because it contains more than one function.
 Tests in these test files do not need grouping into classes, as the file name is used to indicate exactly which function or class is being tested.
+If we are wishing to link tests in R, we can use contexts to combine tests into blocks which test related functionality.
+The following example links two different tests under the same context named "my_sum testing":
+
+```{code-block} r
+context("my_sum testing")
+
+test_that("my_sum returns correct value", {
+    expect_that(my_sum(1,1), equals(2))
+    expect_that(my_sum(-1,1), equals(0))
+})
+
+test_that("my_sum missing value", {
+    expect_that(my_sum(NA,1), equals(NA))
+})
+```
 
 These are the common conventions for each of Python and R, but are interchangeable.
 Use the approach that makes it easiest for developers to identify the relationship between tests and the code they are testing.
 
 Note that some test frameworks allow you to keep the tests in the same file as the code that is being tested.
 This is a good way of keeping tests and code associated,
 but you should ensure that good modular code practices are followed to separate unrelated code into different files.
+Additional arguments are made to separate tests and functions when you are packaging your code.
+If unit tests and code are located together in the same file,
+the unit tests would also be packaged and installed by additional users.
+Therefore when packaging code,
+the unit tests should be moved to an adjacent test folder as users will not need to have unit tests installed when installing the package.
+
+When separating unit tests into main package and testing scripts, its important to import your package to ensure the correct functions are being unit tested.
+For the module structure outlined previously, we would use `from src.math import my_math_function`.
+For R you need to specify the name of your package within the `testthat.R` file within your tests folder.
+
+## Structuring tests
+
+In order to maintain a consistency across modules you develop, you should follow [PEP8](https://www.python.org/dev/peps/pep-0008/) (python)
+or [Google](https://google.github.io/styleguide/Rguide.html) / [tidyverse](https://style.tidyverse.org/) (R) standards when structuring unit tests.
+
+For python this involves importing all needed function at the beginning of the test file.
+To ensure that the correct functions are imported from your module,
+it is also recommend to install a local editable version into your virtual environment.
+This is done by running `pip install -e .` and any changes made to your
+module functions will also be updated in your python environment.
+Following this it is recommend to define fixtures, classes and then test functions.
+An example of this can be found below.
+
+Similar structure should be followed in R, with all modules loaded in the beginning of a test script.
```

Comment:
Can we add a reference to R testing book https://r-pkgs.org/ here? I think it's referenced at the top of the chapter so we could just refer back to that. It would be good if we can do the same for Python too.